documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md 

```diff
@@ -0,0 +1,82 @@
+- As a maintainer, I've found a discrepancy between the type of data property the API returns and what is described in the table (e.g., is `null` for a non-nullable property). Is there an issue to fix this data in the catalog database, or do I need to create a new one?
+- As an API user, I need to know what exactly this media property is and what data points from the providers are used for it. E.g., does the "creator" refer to the person who created the object or took the picture?
+- As a provider of openly-licensed media, I want to know what media properties Openverse collects so that I can change my API responses.
+
+The information on data types, transformations, and validations will be presented as a table and will be extracted from the code.  Other information about the data selection properties and data discrepancies will be written in a separate markdown file as it is easier to read, write, and lint.
+
+The catalog has a DAG documentation generator that extracts the docstrings from the DAGs and creates a page with the documentation. We can use the same approach to create a documentation page for the media properties. The DAG doc generator runs the code inside docker to extract DAG information. In this project, we can simplify and use the `ast` module to parse static code and extract the docstrings to make the checks faster. This will, however, mean that some of the settings will not be picked up and will have to be updated during the generation. One example of this is the `nullable` field in the `Column` class: it falls back to the value of `not nullable` during the object initiation.
+
+## Outlined Steps
+
+### `just` script
+
+The script parses the SQL DDL files, Python files and markdown files to generate a markdown page with a table and long-form notes. This script can `diff` with the existing documentation page (created in step 4 below) for the `precommit` check and post a note about how to update the docs if differences are detected.
+
+### 1. SQL DDL parsing
+We will use the names of the properties as they are in the upstream database, in the same order as they are in the `image_view`/`audio_view` materialized views. These views are selected over the main `image` and `audio` tables because they are used by the API during data refresh and contain a complete list of properties.
+
+A Python script is written to extract the information from the SQL files into a mapping with the name of the media property and the following values: SQL datatype, and SQL constraints. More data will be added to this mapping in the further steps.
+
+### 2. Parsing of the Python code
+The `Column` class and its child classes describe the validations we use to write the data collected by provider scripts. The `add_item` method has docstrings with short descriptions of what each property is. The information from these items will be added into the mapping from step 1.
+
```

Comment:
Does any of the validation in the `MediaStore` need to be captured, or is that all captured in the `Column` definitions? (For example I do think that all the fields checked as required fields in the MediaStore are actually non-nullable)

Reply:
It is very difficult to parse out validation from the `MediaStore` using ast, so I thought that the validation for each property would be manually described in the markdown file.

Do you think that it would be reasonable and useful to track changes in the `MediaStore` (and the `ImageStore` and `AudioStore`) and warn a contributor to update the documentation if necessary? This could be a precommit warning (I don't like this idea because it's very easy to miss; or even if it is possible to add a non-error warning to precommit) or a CI step with a comment on the PR. We cannot use an error instead of a warning for that because the changes to media store might not actually need to have corresponding changes to the documentation.
For implementing this warning, we can use `ast` to get the line numbers for `add_item` and `clean_metadata`, and then check `git diff` to see if any of these lines are affected. I'm not sure what to do about the functions that `clean_metadata` calls, though.

Reply:
Totally, I don't see how those sorts of changes could possibly be parsed.

> Do you think that it would be reasonable and useful to track changes in the MediaStore (and the ImageStore and AudioStore) and warn a contributor to update the documentation if necessary?

Very cool idea! A comment on the PR seems ideal and mirrors migration warnings.

> For implementing this warning, we can use ast to get the line numbers for add_item and clean_metadata, and then check git diff to see if any of these lines are affected. I'm not sure what to do about the functions that clean_metadata calls, though.

It might even make sense to just issue the warning if any of the `MediaStore` files are changed at all, as long as it's a warning and not a blocker. What do you think?

Reply:
I love the idea of a warning when `MediaStore` files are changed, it greatly simplifies the implementation :) And is still very useful.